File: codingame/python/OrganicCompounds.py
import re
import unittest

def getBondNum(bond: str) -> int:
    # get m of "(m)"
    if len(bond.strip()) != 3: return 0
    return int(bond.strip()[1]) 

def checkCarbonAround(allLines: [], carbon: str, carbonRow: int, carbonCol: int) -> bool:
    # check all the bonds around given carbon
    carLine = allLines[carbonRow]
    # get n of "CHn"
    carbonRides = int(carbon[2])

    upBoundM = rightBoundM = downBoundM = leftBoundM = 0
    # up bound
    if (carbonRow > 0 and len(allLines[carbonRow - 1]) >= carbonCol + 3):
        upBoundM = getBondNum(allLines[carbonRow-1][carbonCol : carbonCol+3])
    # right bound
    if (carbonCol + 6 <= len(carLine)):
        rightBoundM = getBondNum(carLine[carbonCol+3 : carbonCol+6])
    # down bound
    if (carbonRow < len(allLines)-1 and len(allLines[carbonRow + 1]) >= carbonCol + 3):
        downBoundM = getBondNum(allLines[carbonRow+1][carbonCol : carbonCol+3])
    # left bound
    if (carbonCol - 3 >= 0):
        leftBoundM = getBondNum(carLine[carbonCol-3 : carbonCol])
    # sum up
    return carbonRides + upBoundM + rightBoundM + downBoundM + leftBoundM == 4

def check_compounds_valid(lines: list[str]) -> bool:
    carbonPat = re.compile(r'(CH[0-4])')
    for i in range(len(lines)):
        line = lines[i]
        beginIdx = 0
        while (mat := carbonPat.search(line, pos=beginIdx)):
            carbonBeginIdx, carbon = mat.start(), mat.group()
            if not checkCarbonAround(lines, carbon, i, carbonBeginIdx):
                return False
            beginIdx = carbonBeginIdx + len(mat.group())
    return True

class TestOrganicCompoundsCheck(unittest.TestCase):
    # each test case must start with 'test_'
    def test_valid(self):
        org_compounds = [
            "CH2(1)CH2(1)CH1(1)CH2",
            "(1)         (1)   (1)",
            "CH2         CH2   CH2",
            "(1)         (1)   (1)",
            "CH2(1)CH2(1)CH1(1)CH2"
        ]
        self.assertTrue(check_compounds_valid(org_compounds))

    def test_invalid(self):
        org_compounds = [
            "CH3(1)CH1(1)CH2(1)CH1(1)CH3",
            "(1)         (1)",
            "CH3         CH1(1)CH1(1)CH3",
            "            (1)",
            "            CH3"
        ]
        self.assertFalse(check_compounds_valid(org_compounds))

if __name__ == '__main__':
    unittest.main()

File: codingame/java/src/main/OrganicCompounds.java
import java.util.*;
import java.util.regex.*;
import java.io.*;
import java.math.*;

class OrganicCompounds {
    // entry
    public boolean checkCompoundsValid(List<String> compLines) {
        Pattern carbonPat = Pattern.compile("CH[0-4]");
        for (int i = 0; i < compLines.size(); i++) {
            Matcher carbonMat = carbonPat.matcher(compLines.get(i));
            while (carbonMat.find()) {
                int bondEnd = carbonMat.end();
                if (!checkCarbon(compLines, i, bondEnd-3)) {
                    return false;
                }
            }
        }
        return true;
    }

    // check carbon: whether all the bounds number around it sum up with carbon number equals to 4 
    private boolean checkCarbon(List<String> allLines, int rowIdx, int beginColumn) {
        String carbon = allLines.get(rowIdx).substring(beginColumn, beginColumn+3);
        int carbonRides = Integer.valueOf(
            carbon.substring(2,3)); // n of "CHn", valid range is [0, 4]
        if (carbonRides < 0 || carbonRides > 4) return false;

        int upBondM = 0, rightBondM = 0, downBondM = 0, leftBondM = 0;
        // up bond
        if (rowIdx > 0 && allLines.get(rowIdx-1).length() >= beginColumn+3) // up line should be long enough
            upBondM = getBondNum(allLines.get(rowIdx-1).substring(beginColumn, beginColumn+3));
        // right bond
        if (beginColumn + 6 <= allLines.get(rowIdx).length())
            rightBondM = getBondNum(allLines.get(rowIdx).substring(beginColumn+3, beginColumn+6));
        // down bound
        if (rowIdx < allLines.size() - 1 && allLines.get(rowIdx+1).length() >= beginColumn+3) // down line should be long enough 
            downBondM = getBondNum(allLines.get(rowIdx+1).substring(beginColumn, beginColumn+3));
        // left bound
        if (beginColumn - 3 >= 0)
            leftBondM = getBondNum(allLines.get(rowIdx).substring(beginColumn-3, beginColumn));
        // check carbon and all bonds round it
        return (carbonRides + upBondM + rightBondM + downBondM + leftBondM) == 4;
    }

    // return 0 in case of invalid bound
    public static int getBondNum(String bond) {
        if (bond.strip().length() != 3) return 0;
        if (!bond.startsWith("(") || !bond.endsWith(")")) return 0;
        int m = Integer.valueOf(bond.substring(1, 2)); // m of "(m)", valid range is (0, 4]
        return (m > 0 && m <= 4) ? m : 0;
    }
}